src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,434 @@
+#include <Core/NamesAndTypes.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/IDataType.h>
+#include <DataTypes/Serializations/ISerialization.h>
+#include <IO/ReadBuffer.h>
+#include <IO/WriteHelpers.h>
+#include <Processors/Formats/IRowInputFormat.h>
+#include <base/types.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_DATA;
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+}
+
+
+DataTypePtr createDataType(size_t depth, DataTypePtr nested_type)
+{
+    DataTypePtr result_type = nested_type;
+
+    assert(depth > 0);
+    if (depth > 1)
+    {
+        for (size_t i = 0; i < depth - 1; ++i)
+            result_type = std::make_shared<DataTypeArray>(std::move(result_type));
+    }
+    return result_type;
+}
+
+/*
+Checks, in what endian format data was written.
+return -1: if data is written in little-endian;
+
+         1: if data is written in big-endian;
+
+         0: if data is written in no-endian. */
+int endianOrientation(String descr)
+{
+    if (descr.length() < 3)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Descr field length must be bigger or equal 3");
+    if (descr[0] == '<')
+        return -1;
+    else if (descr[0] == '>')
+        return 1;
+    else if (descr[0] == '|')
+        return 0;
+    else
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Wrong content of field descr");
+}
+
+DataTypePtr parseType(String type)
+{
+    if (type == "<i1")
+        return std::make_shared<DataTypeInt8>();
+    else if (type == "<i2")
+        return std::make_shared<DataTypeInt16>();
+    else if (type == "<i4")
+        return std::make_shared<DataTypeInt32>();
+    else if (type == "<i8")
+        return std::make_shared<DataTypeInt64>();
+    else if (type == "<u1")
+        return std::make_shared<DataTypeUInt8>();
+    else if (type == "<u2")
+        return std::make_shared<DataTypeUInt16>();
+    else if (type == "<u4")
+        return std::make_shared<DataTypeUInt32>();
+    else if (type == "<u8")
+        return std::make_shared<DataTypeUInt64>();
+    else if (type == "<f2")
+        return std::make_shared<DataTypeFloat32>();
+    else if (type == "<f4")
+        return std::make_shared<DataTypeFloat32>();
+    else if (type == "<f8")
+        return std::make_shared<DataTypeFloat64>();
+    else if (type == "<c8" || type == "<c16")
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support complex numeric type");
+    else if (type == "|b1")
+        return std::make_shared<DataTypeInt8>();
+    else if (type[1] == 'U' || type[1] == 'S')
+        return std::make_shared<DataTypeString>();
+    else if (type == "O")
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support object types");
+    else
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Error while parsing data type");
+}
+
+std::vector<int> parseShape(String shapeString)
```

Comment:
Please, use underscores instead of `cumelCase` for variable names: `shapeString` -> `shape_string`.